File: OrangeJuice.Server.Core.FSharp/Data/Logging/ErrorLogFactory.fs
namespace OrangeJuice.Server.FSharp.Data.Logging

open Elmah
open Elmah.Contrib.EntityFramework

open OrangeJuice.Server.Configuration

type ErrorLogFactory(environmentProvider : IEnvironmentProvider, connectionStringProvider : IConnectionStringProvider) =
    interface Factory.IFactory<ErrorLog> with
        member this.Create() : ErrorLog =
            this.Create()

    member this.Create() =
        match environmentProvider.GetCurrentEnvironment() with
            | Environment.Production -> new EntityErrorLog(connectionStringProvider.GetDefaultConnectionString()) :> ErrorLog
            | _ -> new TraceErrorLog() :> ErrorLog

File: OrangeJuice.Server.Core/Data/Logging/ErrorLogFactory.cs
using Elmah;

using OrangeJuice.Server.Configuration;

namespace OrangeJuice.Server.Data.Logging
{
	public sealed class ErrorLogFactory : Factory.IFactory<ErrorLog>
	{
		private readonly IEnvironmentProvider _environmentProvider;
		private readonly IConnectionStringProvider _connectionStringProvider;

		public ErrorLogFactory(IEnvironmentProvider environmentProvider, IConnectionStringProvider connectionStringProvider)
		{
			_environmentProvider = environmentProvider;
			_connectionStringProvider = connectionStringProvider;
		}

		public ErrorLog Create()
		{
			switch (_environmentProvider.GetCurrentEnvironment())
			{
				case Environment.Production:
					return new Elmah.Contrib.EntityFramework.EntityErrorLog(_connectionStringProvider.GetDefaultConnectionString());
				default:
					return new TraceErrorLog();
			}
		}
	}
}